File: Sqrt.scala
// uses Newton's method
object Sqrt {
    def abs(x: Double) = if (x >= 0) x else -x

    def sqrt(x: Double) = {
        def sqrtIter(guess: Double): Double =
            if (isGoodEnough(guess)) guess 
            else sqrtIter(improve(guess)) 
        def isGoodEnough(guess: Double) =
            abs(guess * guess - x) / x < 0.0001 // note we divide by x so that it works for both large and small numbers

        def improve(guess: Double) =
            (guess + x / guess) / 2
        sqrtIter(1.0)
    }

}

File: Sqrt.java
// uses Newton's method
public class Sqrt {
    public static void compute_sqrt(Integer n) {
      System.out.println((int) sqrt(1, n));
    }

    private static double sqrt(double guess, double x) {
        if (isGoodEnough(guess, x)) return guess; 
        else return sqrt(improve(guess, x), x); 
    }
    private static boolean isGoodEnough(double guess, double x) {
      System.out.println("Comparing " + guess + " with " + x);
      return Math.abs(guess * guess - x) / x < 0.0001; // note we divide by x so that it works for both large and small numbers
    }
    private static double improve(double guess, double x) {
      return (guess + x / guess) / 2;
    }
    public static void main(String[] args) {
      compute_sqrt(17);
    }
}

